xdem/coreg/base.py

```diff
@@ -748,47 +748,223 @@ def invert_matrix(matrix: NDArrayf) -> NDArrayf:
     :param z_name: Column name to use as elevation, only for point elevation data passed as geodataframe.
-    :return:
+
+    :return: Transformed elevation point cloud.
     """
 
-    if isinstance(elev, gpd.GeoDataFrame):
-        return _apply_matrix_pts(epc=elev, matrix=matrix, invert=invert, centroid=centroid, z_name=z_name)
-    else:
-        if isinstance(elev, gu.Raster):
-            transform = elev.transform
-            dem = elev.data
-        else:
-            dem = elev
+    # Apply transformation to X/Y/Z arrays
+    tx, ty, tz = _apply_matrix_pts_arr(
+        x=epc.geometry.x.values,
+        y=epc.geometry.y.values,
+        z=epc[z_name].values,
+        matrix=matrix,
+        centroid=centroid,
+        invert=invert,
+    )
 
-        # TODO: Add exception for translation to update only geotransform, maybe directly in apply_matrix?
-        applied_dem = _apply_matrix_rst(
-            dem=dem, transform=transform, matrix=matrix, invert=invert, centroid=centroid, resampling=resampling
-        )
-        if isinstance(elev, gu.Raster):
-            applied_dem = gu.Raster.from_array(applied_dem, transform, elev.crs, elev.nodata)
-        return applied_dem
+    # Finally, transform back to a new GeoDataFrame
+    transformed_epc = gpd.GeoDataFrame(
+        geometry=gpd.points_from_xy(x=tx, y=ty, crs=epc.crs),
+        data={z_name: tz},
+    )
+
+    return transformed_epc
+
+
+def _iterate_affine_regrid_small_rotations(
+    dem: NDArrayf,
+    transform: rio.transform.Affine,
+    matrix: NDArrayf,
+    centroid: tuple[float, float, float] | None = None,
+    resampling: Literal["nearest", "linear", "cubic", "quintic"] = "linear",
+    verbose: bool = True,
+) -> tuple[NDArrayf, rio.transform.Affine]:
+    """
+    Iterative process to find the best reprojection of affine transformation for small rotations.
+
+    Faster than regridding point cloud by triangulation of points (for instance with scipy.interpolate.griddata).
+    """
+
+    # Convert DEM to elevation point cloud, keeping all exact grid coordinates X/Y even for NaNs
+    dem_rst = gu.Raster.from_array(dem, transform=transform, crs=None, nodata=99999)
+    epc = dem_rst.to_pointcloud(data_column_name="z", skip_nodata=False).ds
+
+    # Exact affine transform of elevation point cloud (which yields irregular coordinates in 2D)
+    tz0 = _apply_matrix_pts_arr(
+        x=epc.geometry.x.values, y=epc.geometry.y.values, z=epc.z.values, matrix=matrix, centroid=centroid
+    )[2]
+
+    # We need to find the elevation Z of a transformed DEM at the exact grid coordinates X,Y
+    # Which means we need to find coordinates X',Y',Z' of the original DEM that, after the exact affine transform,
+    # fall exactly on regular X,Y coordinates
+
+    # 1/ The elevation of the original DEM, Z', is simply a 2D interpolator function of X',Y' (bilinear, typically)
+    xycoords = dem_rst.coords(grid=False)
+    z_interp = scipy.interpolate.RegularGridInterpolator(
```

Comment:
Why not use Raster.interp here?

Reply:
For computational speed: better to create the interpolator only once, and re-use it everywhere, then re-do it at every step!
Maybe we can have this done in GeoUtils directly in the future (when we move the function contents outside of the `Raster` class)

Reply:
Added a comment to specify this